qiskit/transpiler/passes/optimization/optimize_1q_commutation.py

```diff
@@ -0,0 +1,226 @@
+        """
+        Replaces a nonempty sequence `old_run` of `DAGNode`s, assumed to be a complete chain in
+        `dag`, with the circuit `new_circ`.
+        """
+
+        new_dag = circuit_to_dag(new_circ)
+        dag.substitute_node_with_dag(old_run[0], new_dag)
+        for node in old_run[1:]:
+            dag.remove_op_node(node)
+
+    def _step(self, dag):
+        """
+        Performs one unit of optimization work.
+
+        Returns True if `dag` changed, False if no work on `dag` was possible.
+        """
+
+        # NOTE: This next call is somewhat expensive, and it could be improved if
+        #       `dag.substitute_node_with_dag` returned the new (ordered) replacement node list.
+        runs = dag.collect_1q_runs()
+        for run in runs:
+            # identify the preceding blocking gates
+            run_clone = copy(run)
+            if run == []:
+                continue
+
+            # try to modify preceding_run
+            preceding_blocker, preceding_run = self._find_adjoining_run(dag, runs, run)
+            commuted_preceding = []
+            if preceding_run != []:
+                commuted_preceding, run_clone = self._commute_through(preceding_blocker, run_clone)
+
+            # try to modify succeeding run
+            succeeding_blocker, succeeding_run = self._find_adjoining_run(
+                dag, runs, run, front=False
+            )
+            commuted_succeeding = []
+            if succeeding_run != []:
+                run_clone, commuted_succeeding = self._commute_through(
+                    succeeding_blocker, run_clone, front=False
+                )
+
+            # re-synthesize
+            new_preceding_run = self._resynthesize(preceding_run + commuted_preceding)
+            new_succeeding_run = self._resynthesize(commuted_succeeding + succeeding_run)
+            new_run = self._resynthesize(run_clone)
+
+            # check whether this was actually a good idea
+            original_depth = len(run) + len(preceding_run) + len(succeeding_run)
+            new_depth = len(new_run) + len(new_preceding_run) + len(new_succeeding_run)
+
+            # perform the replacement if it was indeed a good idea
+            if original_depth > new_depth:
+                if preceding_run != []:
+                    self._replace_subdag(dag, preceding_run, new_preceding_run)
+                if succeeding_run != []:
+                    self._replace_subdag(dag, succeeding_run, new_succeeding_run)
+                self._replace_subdag(dag, run, new_run)
+                return True
+
+        return False
+
+    def run(self, dag):
+        """
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
+        """
+
+        # python doesn't support tail calls
+        while True:
+            did_work = self._step(dag)
+            if not did_work:
+                break
```

Comment:
Can we keep this pass to performing one unit of optimization (literally "a pass" over the circuit), and delegate this `do-while` loop to the pass manager? I think that is more in the spirit of pass vs. passmanager. 

Something like:

```py
from qiskit.transpiler.passes import Optimize1qGatesSimpleCommutation, Size, FixedPoint

def repeat_condition(property_set):
    return not property_set['size_fixed_point']

check_size = [Size(), FixedPoint('size')]
optimize = [Optimize1qGatesSimpleCommutation()]

pm = PassManager()
pm.append(check_size + optimize, do_while=repeat_condition)

new_circ = pm.run(circ)
```

Reply:
I'm happy to make this change, but I'd warn that it may not do what one expects: `_step` finds a single optimization opportunity, because acting on that opportunity invalidates the output of `collect_1q_runs`, forcing a restart. From the perspective of someone running the pass in a while loop, this difference won't matter; but someone who invokes it "bare" might be surprised at how little work it does.

Reply:
Added this change, but kept the old behavior as a flag.